myokit/formats/diffsl/_exporter.py

```diff
@@ -0,0 +1,523 @@
+    are output, but no inputs are provided.
+
+    For details of the language, see https://martinjrobins.github.io/diffsl/
+    """
+
+    def model(self, path, model, protocol=None, convert_units=True):
+        """
+        Exports a :class:`myokit.Model` in DiffSL format, writing the result to
+        the file indicated by ``path``.
+
+        A :class:`myokit.ExportError` will be raised if any errors occur.
+
+        Arguments:
+
+        ``path``
+            The path to write the generated model to.
+        ``model``
+            The :class:`myokit.Model` to export.
+        ``protocol``
+            Not implemented!
+        ``convert_units``
+            If set to ``True``, the method will attempt to convert to
+            preferred units for voltage (mV), current (A/F), and time (ms).
+        """
+        # Raise exception if path is unwritable
+        self._test_writable_dir(os.path.dirname(path))
+
+        # Check model validity
+        try:
+            model.validate()
+        except myokit.MyokitError as e:
+            raise myokit.ExportError(
+                'DiffSL export requires a valid model.'
+            ) from e
+
+        # Prepare model for export
+        model, currents = self._prep_model(model, convert_units)
+
+        # Generate DiffSL model
+        diffsl_model = self._generate_diffsl(
+            model,
+            extra_out_vars=currents,
+        )
+
+        # Write DiffSL model to file
+        with open(path, 'w') as f:
+            f.write(diffsl_model)
+
+    def _prep_model(self, model, convert_units):
+        """
+        Prepare the model for export to DiffSL.
+        """
+        # Rewrite model so that any Markov models have a 1-sum(...) state
+        # This also clones the model, so that changes can be made
+        model = markov.convert_markov_models_to_compact_form(model)
+
+        # Remove all model bindings apart from time and pace
+        _ = myokit._prepare_bindings(
+            model,
+            {
+                'time': 't',
+                'pace': 'pace',
+            },
+        )
+
+        # Remove hardcoded stimulus protocol, if any
+        guess.remove_embedded_protocol(model)
+
+        # Get / try to guess some model variables
+        time = model.time()  # engine.time
+        vm = self._guess_membrane_potential(model)  # Vm
+        cm = guess.membrane_capacitance(model)  # Cm
+        currents = self._guess_currents(model)
+
+        time_factor = time_factor_inv = None
+        if convert_units:
+            # Convert currents to A/F
+            helpers = [] if cm is None else [cm.rhs()]
+            for var in currents:
+                self._convert_current_unit(var, helpers=helpers)
+
+            # Convert potentials to mV
+            self._convert_potential_unit(vm)
+
+            time_factor, time_factor_inv = self._get_time_factor(time)
+
+        # Find HH state variables and their alphas, betas, and taus
```

Comment:
Don't think you need to do this here. EasyML required HH equations in inf-tau form, but DiffSL just treats them as ODEs so no special treatment required

Reply:
Resolved by 1a21249